samples/wasi-threads/wasm-apps/thread_termination.c

```diff
@@ -27,7 +30,7 @@ typedef struct {
 void
 run_long_task()
 {
-    // Busy waiting to be interruptible by exception
+    // Busy waiting to be interruptible by trap or `proc_exit`
     for (int i = 0; i < TIMEOUT_SECONDS; i++)
```

Comment:
@wenyongh btw I notice something while testing: if I replace this for loop with a `__builtin_wasm_memory_atomic_wait32(0, 0, -1);` or even `while(1);` the thread doesn't get interrupted even if the exception is propagated (in the first case it gets stuck here https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/common/posix/posix_thread.c#L192). Is that expected? I'll investigate to understand what's happening

Reply:
it's what i expect to happen with the current implementation.

to terminate busy looping threads, something like `wasm_cluster_thread_send_signal` would work.
i guess we eventually need to implement something (eg. send a signal on posix-like platforms) to terminate the blocking threads.

Reply:
> @wenyongh btw I notice something while testing: if I replace this for loop with a `__builtin_wasm_memory_atomic_wait32(0, 0, -1);` or even `while(1);` the thread doesn't get interrupted even if the exception is propagated (in the first case it gets stuck here https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/common/posix/posix_thread.c#L192). Is that expected? I'll investigate to understand what's happening

Normally runtime should terminate all threads when exception was thrown, I remember it is the expected behavior of Java VM. Seems it is an issue of the main branch, I try to fix it with the patch below for interpreter. And another issue of main branch may be that the main thread doesn't spread exception to other thread when exception was thrown, like the change of this PR in wasm_runtime_call_wasm. I will discuss with @xujuntwt95329 for how to fix them.
```diff
diff --git a/core/iwasm/interpreter/wasm_interp_fast.c b/core/iwasm/interpreter/wasm_interp_fast.c
index 3109b0c8..36863bf2 100644
--- a/core/iwasm/interpreter/wasm_interp_fast.c
+++ b/core/iwasm/interpreter/wasm_interp_fast.c
@@ -1054,6 +1054,8 @@ wasm_interp_call_func_import(WASMModuleInstance *module_inst,
 #if WASM_ENABLE_THREAD_MGR != 0
 #define CHECK_SUSPEND_FLAGS()                           \
     do {                                                \
+        if (wasm_get_exception(module))                 \
+            goto got_exception;                         \
         if (exec_env->suspend_flags.flags != 0) {       \
             if (exec_env->suspend_flags.flags & 0x01) { \
```

Reply:
@eloparco I discussed the issues with @xujuntwt95329, he agreed that they are issues and will submit patch to fix them on main branch. Could you wait until we fix them and merge branch `main` into `dev/wasi_threads`, so as to sync up the code and avoid conflicts in the future?

Reply:
as proposed by @yamt, I tried adding a `wasm_cluster_send_signal_all(cluster, WAMR_SIG_TERM);` after spreading the exception https://github.com/bytecodealliance/wasm-micro-runtime/blob/903f5212f194c2709b32631bd32936fdcfed27af/core/iwasm/libraries/thread-mgr/thread_manager.c#L995 and it solves in case of a `while(1);` but not for other cases like `sleep(TIMEOUT_SECONDS);` or `__builtin_wasm_memory_atomic_wait32(0, 0, -1);`

Reply:
> but not for other cases like `sleep(TIMEOUT_SECONDS);` or `__builtin_wasm_memory_atomic_wait32(0, 0, -1);`

i guess something more complex is necessary for blocking threads.
like use `pthread_kill` and make them handle `EINTR`.

for another runtime i made all i/o non-blocking to handle thread termination requests. but i don't think it's an appropriate approach for WAMR.

Reply:
@eloparco Yes, `wasm_cluster_send_signal_all` can only tell the interpreter to exit the execution loop, so `while(1)` can be terminated, but for `sleep` and `__builtin_wasm_memory_atomic_wait32`, they are not executed in interpreter, they have already fall into OS kernel, so the `wasm_cluster_send_signal_all` will not work for this situation.

I agree with @yamt that we may need some more complex mechanism to terminate the threads when they are blocked in kernel.

Reply:
Is that also a problem in lib_pthread? I wonder if we should fix it in the `main` branch as well.

Reply:
> Is that also a problem in lib_pthread?

yes; it applies in general to threads running "blocking" instructions like `sleep()` or `__builtin_wasm_memory_atomic_wait32()`

> I wonder if we should fix it in the main branch as well.

I think that's what @wenyongh was proposing a few comments above

Reply:
@loganek @eloparco Jun's PR #1889 was merged and I tried to create a PR to merge branch main into dev/wasi_threads, but unfortunately there are conflicts in below files:
```
        .github/workflows/compilation_on_android_ubuntu.yml
        core/iwasm/libraries/thread-mgr/thread_manager.c
        core/iwasm/libraries/thread-mgr/thread_manager.h
```
I can resolve these conflicts, and the diff file of the PR is as below:
[merge_main_into_wasi_threads.zip](https://github.com/bytecodealliance/wasm-micro-runtime/files/10462938/merge_main_into_wasi_threads.zip)

Could you help check whether the patch works for you? If yes, I will push it manually under the command line. Thanks.

Reply:
Hi @wenyongh , I reviewed the patch and it is mostly correct, except one thing. In function `thread_manager_start_routine()` you moved `free_aux_stack` below `wasm_runtime_deinstantiate_internal`; `free_aux_stack` must be executed first because it uses module instance. 

Also, one minor thing about `wasm_cluster_add_exec_env` - the `ret` variable is no longer needed so we could remove it; but that's not a big deal.

You can either update your merge commit or feel free to merge it as it is, and I'll open a PR to address the problem.

Reply:
@loganek Thanks for the review, I changed the wasm_runtime_deinstantiate_internal in thread_manager_start_routine. But for the wasm_cluster_add_exec_env, not sure why `ret` can be removed.
I pushed the PR, see: https://github.com/bytecodealliance/wasm-micro-runtime/commit/c7141894fb7b5ef5d39f13748382c9138ee945de

Reply:
> But for the wasm_cluster_add_exec_env, not sure why ret can be removed.

It can be replaced with `return` statements (we no longer have a lock to unlock). As I said, it's not a big deal though. The rest looks ok for me.

Reply:
> I agree with @yamt that we may need some more complex mechanism to terminate the threads when they are blocked in kernel.

@xujuntwt95329 / @wenyongh does anybody on your side actively working on that? If not, we'll take on that work.

Reply:
@loganek not yet